### PR TITLE
nixos/nginx: validate config at build time

### PR DESCRIPTION
**Pull request by symphorien**
###### Description of changes

Shamelessly stolen from nixcloud-webservices:
https://github.com/nixcloud/nixcloud-webservices/blob/master/modules/web/webserver/lib/nginx_check_config.nix

The nixos test testing the behavior of nginx in case of faulty config
would not build with this change (on purpose), so I modified it so that
the failure is not syntactic.

cc @qknight 
cc nginx maintainers @thoughtpolice @fpletz @globin @ajs124 @7c6f434c 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by winterqt** [comment]
@ofborg test nginx

**Comment by ajs124** [comment]
Aren't we already sort of doing this here? https://github.com/NixOS/nixpkgs/blob/26a411b2cb177f88856dfbd1c85367a7647d4d61/nixos/modules/services/web-servers/nginx/default.nix#L93

**Comment by symphorien** [comment]
This runs gixy, which is a linter for some specific kinds of semantic issues, but not syntax errors. Note that this change caught a (deliberate) syntax error in the nginx nixos test that was not caught by gixy.

**Review by ajs124 — APPROVED** [review]

**Comment by symphorien** [comment]
Will self-merge in a few days if nobody opposes.

**Review by RaitoBezarius — APPROVED** [review]

**Comment by delroth** [comment]
This breaks in presence of backend DNS names in nginx configuration, since the sandbox can't access DNS:

```
2022/12/22 04:39:50 [emerg] 7#7: host not found in upstream "xxx.yyy.zzz" in /build/conf:793
```

```
proxy_pass https://xxx.yyy.zzz:443;
```

**Comment by RaitoBezarius** [comment]
> This breaks in presence of backend DNS names in nginx configuration, since the sandbox can't access DNS:
> 
> ```
> 2022/12/22 04:39:50 [emerg] 7#7: host not found in upstream "xxx.yyy.zzz" in /build/conf:793
> ```
> 
> ```
> proxy_pass https://xxx.yyy.zzz:443;
> ```

Do you think it's a dealbreaker even if `services.nginx.validateConfig = false;` would disable this feature?

**Comment by symphorien** [comment]
Could you share more of the config? I'm trying to reproduce with the nginx nixos test:
```diff
diff --git a/nixos/tests/nginx.nix b/nixos/tests/nginx.nix
index d9d073822a1..1b06759c162 100644
--- a/nixos/tests/nginx.nix
+++ b/nixos/tests/nginx.nix
@@ -13,6 +13,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
   nodes = {
     webserver = { pkgs, lib, ... }: {
       services.nginx.enable = true;
+      services.nginx.upstreams.yay.servers."1.2.3.4:3000" = {};
       services.nginx.commonHttpConfig = ''
         log_format ceeformat '@cee: {"status":"$status",'
           '"request_time":$request_time,'
@@ -57,6 +58,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
 
       specialisation.reloadRestartSystem.configuration = {
         services.nginx.package = pkgs.nginxMainline;
+        services.nginx.virtualHosts."yay.com".locations."/".proxyPass = "https://1.2.3.4:443";
       };
 
       specialisation.reloadWithErrorsSystem.configuration = {
```
but it still builds.

**Comment by RaitoBezarius** [comment]
> Could you share more of the config? I'm trying to reproduce with the nginx nixos test:
> 
> ```diff
> diff --git a/nixos/tests/nginx.nix b/nixos/tests/nginx.nix
> index d9d073822a1..1b06759c162 100644
> --- a/nixos/tests/nginx.nix
> +++ b/nixos/tests/nginx.nix
> @@ -13,6 +13,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
>    nodes = {
>      webserver = { pkgs, lib, ... }: {
>        services.nginx.enable = true;
> +      services.nginx.upstreams.yay.servers."1.2.3.4:3000" = {};
>        services.nginx.commonHttpConfig = ''
>          log_format ceeformat '@cee: {"status":"$status",'
>            '"request_time":$request_time,'
> @@ -57,6 +58,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
>  
>        specialisation.reloadRestartSystem.configuration = {
>          services.nginx.package = pkgs.nginxMainline;
> +        services.nginx.virtualHosts."yay.com".locations."/".proxyPass = "https://1.2.3.4:443";
>        };
>  
>        specialisation.reloadWithErrorsSystem.configuration = {
> ```
> 
> but it still builds.

You need DNS in your `proxyPass`, try something like `https://example.com:443` and maybe some DNSMasq pointing `example.com` to `1.2.3.4`.

**Comment by symphorien** [comment]
```diff
diff --git a/nixos/tests/nginx.nix b/nixos/tests/nginx.nix
index d9d073822a1..4172f3ea365 100644
--- a/nixos/tests/nginx.nix
+++ b/nixos/tests/nginx.nix
@@ -13,6 +13,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
   nodes = {
     webserver = { pkgs, lib, ... }: {
       services.nginx.enable = true;
+      services.nginx.upstreams.yay.servers."example.com:3000" = {};
       services.nginx.commonHttpConfig = ''
         log_format ceeformat '@cee: {"status":"$status",'
           '"request_time":$request_time,'
@@ -57,6 +58,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
 
       specialisation.reloadRestartSystem.configuration = {
         services.nginx.package = pkgs.nginxMainline;
+        services.nginx.virtualHosts."yay.com".locations."/".proxyPass = "https://some.dns.info:443";
       };
 
       specialisation.reloadWithErrorsSystem.configuration = {
```
also builds sucessfully.

**Comment by delroth** [comment]
```
diff --git a/nixos/tests/nginx.nix b/nixos/tests/nginx.nix
index 73f1133bd6c..b1010ac1f7d 100644
--- a/nixos/tests/nginx.nix
+++ b/nixos/tests/nginx.nix
@@ -30,6 +30,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
         extraConfig = ''
           access_log syslog:server=unix:/dev/log,facility=user,tag=mytag,severity=info ceeformat;
           location /favicon.ico { allow all; access_log off; log_not_found off; }
+          location / { proxy_pass https://foo.bar.com:443; }
         '';
       };

```

```
nginx config validation failed.
config was /nix/store/mkhs81aygpwhhc44q6iispfy9lvsdswa-nginx.conf.
in case of false positive, set `services.nginx.validateConfig` to false.
nginx output:
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (2: No such file or directory)
2022/12/22 22:07:17 [emerg] 7#7: host not found in upstream "foo.bar.com" in /build/conf:44
nginx: configuration file /build/conf test failed
```

**Comment by symphorien** [comment]
forget what I said, I was not trying on a recent enough revision, it did not contain this PR yet...

I can reproduce, thanks

**Comment by winterqt** [comment]
Perhaps we just disable this by default, or document that one should disable it in this case?

**Comment by symphorien** [comment]
I propose a fix in https://github.com/NixOS/nixpkgs/pull/207413

**Comment by Mic92** [comment]
Even with there are still other issues:

```
validated-nginx.conf> nginx: configuration file /etc/nginx.conf test failed
error: build of '/nix/store/95x32m0621fxbpkwn21z5h8zw85k7d1x-validated-nginx.conf.drv' on 'ssh-ng://nix@ryan.dse.in.tum.de' failed: builder for '/nix/store/95x32m0621fxbpkwn21z5h8zw85k7d1x-validated-nginx.conf.drv' failed with exit code 1;
       last 8 log lines:
       > nginx config validation failed.
       > config was /nix/store/p68y009kly52wrkamjwsv58msp6cy980-nginx.conf.
       > in case of false positive, set `services.nginx.validateConfig` to false.
       > nginx output:
       > /nix/store/mgn3zckncarpmswhv9vn57i43f0rn2jr-nginxWithResolver: line 3: cannot redirect standard input from /dev/null: No such file or directory
       > 8::1            foo.bar
       > nginx: [emerg] BIO_new_file("/var/lib/dhparams/nginx.pem") failed (SSL: error:80000002:system library::No such file or directory:calling fopen(/var/lib/dhparams/nginx.pem, r) error:10000080:BIO routines::no such file)
       > nginx: configuration file /etc/nginx.conf test failed
       For full logs, run 'nix log /nix/store/95x32m0621fxbpkwn21z5h8zw85k7d1x-validated-nginx.conf.drv'.
```

**Comment by Mic92** [comment]
Please merge https://github.com/NixOS/nixpkgs/compare/master...Mic92:nixpkgs:nginx-fix?expand=1 for now so that we can review potential fixes to this problem without hurry.

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 1]
```diff
@@ -489,6 +521,15 @@ in
         '';
       };
 
+      validateConfig = mkOption {
+        default = pkgs.stdenv.hostPlatform == pkgs.stdenv.buildPlatform;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:525]: This should be `pkgs.´stdenv.buildPlatform.canExecute stdenv.hostPlatform``

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 2]
```diff
@@ -391,6 +391,38 @@ let
   );
 
   mkCertOwnershipAssertion = import ../../../security/acme/mk-cert-ownership-assertion.nix;
+
+  snakeOilCert = pkgs.runCommand "nginx-config-validate-cert" { nativeBuildInputs = [ pkgs.openssl.bin ]; } ''
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:395]: Not sure if we want to use runCommandLocal here. We probably don't want to cache or substitute this, or do we?

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 3]
```diff
@@ -391,6 +391,38 @@ let
...
+    openssl req -new -key $out/server.key -out server.csr \
+    -subj "/C=UK/ST=Warwickshire/L=Leamington/O=OrgName/OU=IT Department/CN=example.com"
+    openssl x509 -req -days 1 -in server.csr -signkey $out/server.key -out $out/server.crt
+  '';
+  validatedConfigFile = pkgs.runCommand "validated-nginx.conf" { nativeBuildInputs = [ cfg.package ]; } ''
+    # nginx absolutely wants to read the certificates even when told to only validate config, so let's provide fake certs
+    sed ${configFile} \
+    -e "s|ssl_certificate .*;|ssl_certificate ${snakeOilCert}/server.crt;|g" \
+    -e "s|ssl_trusted_certificate .*;|ssl_trusted_certificate ${snakeOilCert}/server.crt;|g" \
+    -e "s|ssl_certificate_key .*;|ssl_certificate_key ${snakeOilCert}/server.key;|g" \
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:408]: I am not sure how robust the regex is. For using the nix options it is certainly enough.

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 2 above)
Comment (symphorien) [reply nixos/modules/services/web-servers/nginx/default.nix:395]: this certificate is only used during build, so reusing it or sharing it via a binary cache is not a problem

**Pull request merged**